src/v/archival/archiver_operations_api.h

```diff
@@ -0,0 +1,252 @@
+};
+
+struct upload_results_list {
+    model::ntp ntp;
+    std::deque<std::optional<cloud_storage::segment_record_stats>> stats;
+    std::deque<cloud_storage::upload_result> results;
+    std::deque<cloud_storage::segment_meta> metadata;
+
+    // Insync offset of the uploaded manifest
+    model::offset manifest_clean_offset;
+    model::offset read_write_fence;
+
+    size_t num_put_requests{0};
+    size_t num_bytes_sent{0};
+
+    upload_results_list() = default;
+
+    upload_results_list(
+      model::ntp ntp,
+      std::deque<std::optional<cloud_storage::segment_record_stats>> stats,
+      std::deque<cloud_storage::upload_result> results,
+      std::deque<cloud_storage::segment_meta> metadata,
+      model::offset manifest_clean_offset,
+      model::offset read_write_fence,
+      size_t num_put_requests,
+      size_t num_bytes_sent)
+      : ntp(std::move(ntp))
+      , stats(std::move(stats))
+      , results(std::move(results))
+      , metadata(std::move(metadata))
+      , manifest_clean_offset(manifest_clean_offset)
+      , read_write_fence(read_write_fence)
+      , num_put_requests(num_put_requests)
+      , num_bytes_sent(num_bytes_sent) {}
+
+    upload_results_list(const upload_results_list&) = delete;
+    upload_results_list(upload_results_list&&) noexcept = default;
+    upload_results_list& operator=(const upload_results_list&) = delete;
+    upload_results_list& operator=(upload_results_list&&) noexcept = default;
+
+    bool operator==(const upload_results_list& o) const noexcept;
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const upload_results_list& s);
+};
+
+struct admit_uploads_result {
+    model::ntp ntp;
+    size_t num_succeeded{0};
+    size_t num_failed{0};
+
+    // The in-sync offset of the manifest after update
+    model::offset manifest_dirty_offset;
+
+    bool operator==(const admit_uploads_result&) const = default;
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const admit_uploads_result& s);
+};
+
+/// The API provides operations that can be used by different
+/// upload workflows. The object is not specific to any particular
+/// NTP. Multiple workflows can use single API instance.
+class archiver_operations_api {
+public:
+    archiver_operations_api() = default;
+    archiver_operations_api(const archiver_operations_api&) = delete;
+    archiver_operations_api(archiver_operations_api&&) noexcept = delete;
+    archiver_operations_api& operator=(const archiver_operations_api&) = delete;
+    archiver_operations_api& operator=(archiver_operations_api&&) noexcept
+      = delete;
+    virtual ~archiver_operations_api();
+
+    // Segments
+public:
+    /// Return upload candidate(s) if data is available or nullopt
+    /// if there is not enough data to start an upload.
+    virtual ss::future<result<reconciled_upload_candidates_list>>
+    find_upload_candidates(
+      retry_chain_node&, upload_candidate_search_parameters) noexcept
+      = 0;
```

Comment:
Thanks for splitting these out! Just thinking about other ways that would've made sense to split the big PR: I think it would have made sense to introduce these one method interface at a time. One observation is that it seems like most of these don't really need to be grouped together in the same class, and instead these could've been decoupled from each other and orchestrated by the `ntp_archiver`. If it weren't too much work, I'd ask to consider defining much narrower interfaces, one per method, that could be individually tested more easily without bringing in so many other dependencies.

It's not too hard to envision a sequence of PRs like:
- Introduce `find_upload_candidates`, that takes a `storage::log` and some `upload_candidate_search_parameters` and returns a `reconciled_upload_candidates_list`. In isolation, this is easy to test this because it basically just exercises the Kafka batch read path, and some metadata accounting. This would also be reusable by other cloud storage log implementations.
- Introduce `schedule_uploads`, that just exercises interaction with remote storage and materializing bytes from the local storage stream. Tests might examine `s3_imposter`, and validate aspects of the various metadata (paths are all correct, the manifests doesn't include the current segments, etc).
- Introduce `admit_uploads`, that focuses more on updating the archival metadata stm. AFAICT, this is the main piece that has a strong dependency on cluster/archival. Testing here could've looked like supplying `upload_results_list` of incorrect term, with bad read-write fences, changing leaders before replicating, etc.
- Then `upload_manifest`, which seems like a bit of a mix of `schedule_uploads` and maybe `admin_uploads`.

Anyway, I know you have a large backlog of PRs to pull out of The Big One, and I'm sure it's a ton of work to split them out already, so no pressure to follow this breakdown. But I do request you consider the testing I mentioned in each split, as you introduce implementations of each method.

Reply:
>Also, the implementations will be essentially the same object because they will have the same state under the hood and will depend on each other.

I don't think this would be true if split out. Only one method depends on a storage::log. Only two methods depend on having access to Raft. One only method requires a stream of bytes. The requirements of each method are pretty different and it's true that you can plug in a full cluster::partition (or mock) and test all of these together, but in my mind that makes it generally clunkier to work with and extend.

>The idea was to have the whole process described in one place.

I agree it's important to understand at a high level the entire process, what I would've envisioned is something like a single class that orchestrates calls to the various methods (this will likely be similar to your impl), but the implementations of each method being separate

>Each method is tested independently already so I don't see much value in this.

Fair point. Especially since I bet a lot of your branch depends on this, as I mentioned, I'm okay with leaving this as is.

Reply:
> I don't think this would be true if split out. Only one method depends on a storage::log. Only two methods depend on having access to Raft. One only method requires a stream of bytes. The requirements of each method are pretty different and it's true that you can plug in a full cluster::partition (or mock) and test all of these together, but in my mind that makes it generally clunkier to work with and extend.

This will mean additional orchestration in the ntp_archiver. Can be solved by adding abstraction layer but this abstraction layer will depend on wrappers that `archiver_operations_impl` consumes. I wanted to have one method `make_archiver_operations_api` that takes `remote` and  `partition_manager` and  returns a functioning implementation of the interface that can be consumed by `ntp_archiver`. Of cause I can add such method for every object and then add something else that glues everything together. But the result will be functionally equal to what I have now.

The `ntp_archiver` also consumes the scheduler (which is another new component). Don't want to multiply the deps.